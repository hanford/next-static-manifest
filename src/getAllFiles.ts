import * as fs from 'fs';
import * as path from 'path';

export function getAllFiles(dirPath: string, arrayOfFiles: Array<string> = []) {
  const files = fs.readdirSync(dirPath);

  arrayOfFiles = arrayOfFiles || [];

  files.forEach(file => {
    if (fs.statSync(dirPath + '/' + file).isDirectory()) {
      arrayOfFiles = getAllFiles(dirPath + '/' + file, arrayOfFiles);
    } else {
      if (file === '.DS_Store') return;

      arrayOfFiles.push(path.join(dirPath, '/', file));
    }
  });

  return arrayOfFiles;
}
