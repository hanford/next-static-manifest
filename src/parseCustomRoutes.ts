import findUp from 'find-up';
import { CONFIG_FILE } from 'next/constants';
import loadCustomRoutes, {
  normalizeRouteRegex,
} from 'next/dist/lib/load-custom-routes';
import { defaultConfig } from 'next/dist/server/config-shared';
import { pathToRegexp } from 'path-to-regexp';

// import { RouteEntry } from './utils';

function createMap(route: { source: string; destination: string }) {
  const routeRegex = pathToRegexp(route.source, [], {
    strict: true,
    sensitive: false,
    delimiter: '/', // default is `/#?`, but Next does not pass query info
  });

  let regexSource = routeRegex.source;

  return {
    src: route.source,
    dest: route.destination,
    regex: normalizeRouteRegex(regexSource),
    dynamic: true,
  };
}

export default async function parseCustomRoutes(dir: string) {
  const path = await findUp(CONFIG_FILE, { cwd: dir });

  if (path?.length) {
    const userConfigModule = require(path);

    const userConfig = { ...defaultConfig, ...userConfigModule };

    const { rewrites: _rewrites } = await loadCustomRoutes(userConfig);

    const rewrites = {
      beforeFiles: _rewrites.beforeFiles.map(createMap),
      afterFiles: _rewrites.afterFiles.map(createMap),
    };

    return rewrites;
  }

  return { beforeFiles: [], afterFiles: [] };
}
