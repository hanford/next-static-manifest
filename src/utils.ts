export interface RouteEntry {
  dynamic: boolean;
  src: string;
  dest: string;
  regex: RegExp;
}

const REGEX_KEY = '__REGEXP__:';

export function encode(routes: Array<RouteEntry>) {
  return JSON.stringify(routes, serialize);
}

export function decode(object: any) {
  return JSON.parse(object, deserialize);
}

export function deserialize(_k: string, value: string) {
  if (value.toString().includes(REGEX_KEY)) {
    const m = value.split(REGEX_KEY)[1].match(/\/(.*)\/(.*)?/);

    if (!m) return value;

    return new RegExp(m[1], m[2] || '');
  } else {
    return value;
  }
}

function serialize(_: string, value: string | RegExp) {
  if (value instanceof RegExp) {
    return REGEX_KEY + value.toString();
  } else {
    return value;
  }
}
