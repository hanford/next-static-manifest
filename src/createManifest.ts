import { join } from 'path';
import { cwd } from 'process';
import {
  getRouteRegex,
  isDynamicRoute,
} from 'next/dist/shared/lib/router/utils';

import { RouteEntry } from './utils';
import { getAllFiles } from './getAllFiles';

const html = '.html';
const dynamicRouteRegex = /\/\[[^/]+?\](?=\/|$)/g;

export default function createManifest(dirPath: string) {
  const pagesPath = dirPath || join(cwd(), 'out');
  const pages = getAllFiles(pagesPath);

  const routes = [
    ...pages
      .map((route): RouteEntry | null => {
        const dest = route.replace(pagesPath, '');

        if (!dest.endsWith(html)) return null;

        const src = dest.replace(html, '');
        const { re: regex } = getRouteRegex(src);

        return {
          src,
          dest,
          regex,
          dynamic: isDynamicRoute(src),
        };
      })
      .filter(Boolean),
    {
      dynamic: false,
      regex: getRouteRegex('/').re,
      src: '/',
      dest: '/index.html',
    },
  ];

  // Sort first sorts by whether or not the RouteEntry is dynamic
  // and additionally sorts by RouteEntry.src.length
  const sorted = routes.sort((a, b) => {
    const first = a as RouteEntry;
    const second = b as RouteEntry;

    // if both routes are dynamic, sort by amount of dynamic segments
    if (first.dynamic && first.dynamic === second.dynamic) {
      const firstAmountSegments = (first.src.match(dynamicRouteRegex) || [])
        .length;
      const secondAmountSegments = (second.src.match(dynamicRouteRegex) || [])
        .length;

      // if they don't have the same amount of dynamic segments
      if (firstAmountSegments !== secondAmountSegments) {
        return firstAmountSegments > secondAmountSegments ? 1 : -1;
      }
    }

    // if both have the same amount of dynamic segments OR aren't dynamic,
    // sort by length of 'src'
    if (first.dynamic === second.dynamic) {
      return second.src.length - first.src.length;
    }

    // Otherwise, move dynamic down and exact up
    if (first.dynamic) {
      return 1;
    } else {
      return -1;
    }
  });

  if (sorted.some(k => k?.src === '/[...slug]')) {
    sorted.push(
      ...sorted.splice(
        sorted.findIndex(v => v?.src === '/[...slug]'),
        1
      )
    );
  }

  return sorted as Array<RouteEntry>;
}
