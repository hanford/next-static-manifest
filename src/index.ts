import fs from 'fs';
import { join } from 'path';
import { argv } from 'yargs';

import createManifest from './createManifest';
import { encode } from './utils';

const distDir = (argv.distDir as string) || 'out';
const fileName = (argv.fileName as string) || 'next-static-manifest';

const entryManifest = join(distDir, `${fileName}.json`);

const routes = createManifest(distDir);

fs.writeFile(entryManifest, encode(routes), function(err) {
  if (err) throw err;

  console.log('Next.js static manifest created');
  return;
});
