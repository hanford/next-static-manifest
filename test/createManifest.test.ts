import { join } from 'path';
import createManifest from '../src/createManifest';
import { encode, decode, RouteEntry } from '../src/utils';

const out = join(__dirname, 'out');

describe('create manifest scans ./out directory and creates manifest', () => {
  it('returns manifest', () => {
    expect(createManifest(out)).toMatchSnapshot();
  });

  it('encodes manifest', () => {
    const encoded = encode(createManifest(out));

    expect(encoded).toMatchSnapshot();
  });

  it('decodes manifest', () => {
    const decoded = decode(encode(createManifest(out)));

    expect(decoded).toMatchSnapshot();
  });

  it('encoded decoded regex works', () => {
    const decoded = decode(encode(createManifest(out)));
    const postUrl = '/blog/posts/420';

    expect(decoded.findIndex((v: RouteEntry) => postUrl.match(v.regex))).toBe(
      5
    );
  });
});
