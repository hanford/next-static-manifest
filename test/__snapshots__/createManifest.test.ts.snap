// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create manifest scans ./out directory and creates manifest decodes manifest 1`] = `
Array [
  Object {
    "dest": "/blog/posts/[id]/[comment].html",
    "dynamic": true,
    "regex": /\\^\\\\/blog\\\\/posts\\\\/\\(\\[\\^/\\]\\+\\?\\)\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/blog/posts/[id]/[comment]",
  },
  Object {
    "dest": "/blog/posts/[id].html",
    "dynamic": true,
    "regex": /\\^\\\\/blog\\\\/posts\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/blog/posts/[id]",
  },
  Object {
    "dest": "/index.html",
    "dynamic": false,
    "regex": /\\^\\\\/index\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/index",
  },
  Object {
    "dest": "/index.html",
    "dynamic": false,
    "regex": /\\^\\\\/\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/",
  },
  Object {
    "dest": "/[...slug].html",
    "dynamic": true,
    "regex": /\\^\\\\/\\(\\.\\+\\?\\)\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/[...slug]",
  },
]
`;

exports[`create manifest scans ./out directory and creates manifest encodes manifest 1`] = `"[{\\"src\\":\\"/blog/posts/[id]/[comment]\\",\\"dest\\":\\"/blog/posts/[id]/[comment].html\\",\\"regex\\":\\"__REGEXP__:/^\\\\\\\\/blog\\\\\\\\/posts\\\\\\\\/([^/]+?)\\\\\\\\/([^/]+?)(?:\\\\\\\\/)?$/\\",\\"dynamic\\":true},{\\"src\\":\\"/blog/posts/[id]\\",\\"dest\\":\\"/blog/posts/[id].html\\",\\"regex\\":\\"__REGEXP__:/^\\\\\\\\/blog\\\\\\\\/posts\\\\\\\\/([^/]+?)(?:\\\\\\\\/)?$/\\",\\"dynamic\\":true},{\\"src\\":\\"/index\\",\\"dest\\":\\"/index.html\\",\\"regex\\":\\"__REGEXP__:/^\\\\\\\\/index(?:\\\\\\\\/)?$/\\",\\"dynamic\\":false},{\\"dynamic\\":false,\\"regex\\":\\"__REGEXP__:/^\\\\\\\\/(?:\\\\\\\\/)?$/\\",\\"src\\":\\"/\\",\\"dest\\":\\"/index.html\\"},{\\"src\\":\\"/[...slug]\\",\\"dest\\":\\"/[...slug].html\\",\\"regex\\":\\"__REGEXP__:/^\\\\\\\\/(.+?)(?:\\\\\\\\/)?$/\\",\\"dynamic\\":true}]"`;

exports[`create manifest scans ./out directory and creates manifest returns manifest 1`] = `
Array [
  Object {
    "dest": "/blog/posts/[id]/[comment].html",
    "dynamic": true,
    "regex": /\\^\\\\/blog\\\\/posts\\\\/\\(\\[\\^/\\]\\+\\?\\)\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/blog/posts/[id]/[comment]",
  },
  Object {
    "dest": "/blog/posts/[id].html",
    "dynamic": true,
    "regex": /\\^\\\\/blog\\\\/posts\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/blog/posts/[id]",
  },
  Object {
    "dest": "/index.html",
    "dynamic": false,
    "regex": /\\^\\\\/index\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/index",
  },
  Object {
    "dest": "/index.html",
    "dynamic": false,
    "regex": /\\^\\\\/\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/",
  },
  Object {
    "dest": "/[...slug].html",
    "dynamic": true,
    "regex": /\\^\\\\/\\(\\.\\+\\?\\)\\(\\?:\\\\/\\)\\?\\$/,
    "src": "/[...slug]",
  },
]
`;
